### PR TITLE
Refactor miscellaneous methods from collector language interface

### PR DESCRIPTION
**Pull request by ktbriggs**
Refactor MM_CollectorLanguageInterface methods not explicitly bound to
GC component into existing GC component delegate classes
 
void writeBarrier()
  ->MM_Configuration::writeBarrier()
    in Configuration.hpp
bool collectorHeapRegionDescriptorInitialize()
  ->MM_ConfigurationDelegate::initializeHeapRegionDescriptorExtension()
    in <glue>/ConfigurationDelegate.hpp
void collectorHeapRegionDescriptorTearDown()
  ->MM_ConfigurationDelegate::teardownHeapRegionDescriptorExtension()
    in <glue>/ConfigurationDelegate.hpp
	
void flushNonAllocationCaches()
  ->MM_EnvironmentDelegate::flushNonAllocationCaches() 
    in <glue>/EnvironmentDelegate.hpp
OMR_VMThread *attachVMThread()
  ->MM_EnvironmentDelegate::attachVMThread()
    in <glue>/EnvironmentDelegate.hpp
void detachVMThread()
  ->MM_EnvironmentDelegate::detachVMThread()
    in <glue>/EnvironmentDelegate.hpp
bool objectAllocationNotify()
  ->MM_EnvironmentDelegate::objectAllocationNotify()
    in <glue>/EnvironmentDelegate.hpp
	
void workPacketOverflow_overflowItem()
  ->MM_MarkingDelegate::handleWorkPacketOverflowItem()
    in <glue>/MarkingDelegate.hpp

MM_MarkingDelegate was not available when MM_EnvironmentDelegate was
introduced. The following methods were introduced at that time:

MM_EnvironmentBase::markingPhaseStarted()
MM_EnvironmentBase::markingPhaseFinished()
MM_EnvironmentDelegate::markingStarted()
MM_EnvironmentDelegate::markingFinished()

These methods have been refactored into MM_MarkingScheme and 
MM_MarkingDelegate:

MM_MarkingScheme::workerSetupForGC()
MM_MarkingScheme::workerCleanupAfterGC()
MM_MarkingDelegate::workerSetupForGC()
MM_MarkingDelegate::workerCompleteGC()
MM_MarkingDelegate::workerCleanupAfterGC()

Copyright dates adjusted for all files in commit.

Issue #1250
Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

### CODE REVIEW

**Comment by ktbriggs** [comment]
Good to go.

File: example/glue/EnvironmentDelegate.cpp [hunk 1]
```diff
@@ -21,9 +21,40 @@
...
+	if (NULL != omrVMThread) {
+		OMR_Glue_UnbindCurrentThread(omrVMThread);
+	}
+}
+
+void
+MM_EnvironmentDelegate::flushNonAllocationCaches()
+{
+#if defined(OMR_GC_MODRON_SCAVENGER)
+	MM_EnvironmentStandard *envStd = MM_EnvironmentStandard::getEnvironment(_env);
```
Comment (charliegracie) [inline example/glue/EnvironmentDelegate.cpp]: Every language which uses the scavenger needs this.  It should be in OMR and not in the language delegate I believe.

File: example/glue/EnvironmentDelegate.cpp [hunk 2]
```diff
@@ -21,9 +21,40 @@
...
 #include "EnvironmentDelegate.hpp"
 #include "GCExtensionsBase.hpp"
+#include "SublistFragment.hpp"
+
+OMR_VMThread *
+MM_EnvironmentDelegate::attachVMThread(OMR_VM *omrVM, const char *threadName, uintptr_t reason)
+{
+	OMR_VMThread *omrVMThread = NULL;
+	omr_error_t rc = OMR_ERROR_NONE;
+
```
Comment (charliegracie) [inline example/glue/EnvironmentDelegate.cpp]: Is this code that belongs in OMR and not in a language delegate?

Comment (ktbriggs) [inline example/glue/EnvironmentDelegate.cpp]: I tend to agree, but there are Java-specific concerns that require delegation. We can discuss this as I think we may be approaching a time where we need to look at moving some of the low-level threading support from Java into OMR..

File: example/glue/EnvironmentDelegate.cpp (hunk 1 above)
Comment (ktbriggs) [inline example/glue/EnvironmentDelegate.cpp]: OMR does not know what "non-allocation" caches the language may have attached to each thread. This method is required to permit the language to manage any such local caches during GC.

Comment (charliegracie) [inline example/glue/EnvironmentDelegate.cpp]: I agree with you that languages may have caches they need to flush and this is the right API I do not believe the scavenger rs caches should be a language requirement.  It should be part of OMR

Comment (ktbriggs) [inline example/glue/EnvironmentDelegate.cpp]: I removed all language concerns relating to scavenger remembered set in delegate implementations of flushAllocationCaches(). 

See 2565cee for changes.

**Comment by ktbriggs** [comment]
See 2565cee for changes relating to "outdated" comment above.

File: example/glue/MarkingDelegate.hpp [hunk 3]
```diff
@@ -124,44 +172,31 @@ class MM_MarkingDelegate
...
+	 * primariliy by OMR but such objects are presented here in case additional language-dependent handling is
+	 * required.
+	 *
+	 * This method is informational, no specific action is required.
 	 *
 	 * @param env The environment for the calling thread
 	 */
-	MMINLINE void completeMarking(MM_EnvironmentBase *env) { }
-
+	MMINLINE void handleWorkPacketOverflowItem(MM_EnvironmentBase *env, omrobjectptr_t objectPtr) { }
```
Comment (charliegracie) [inline example/glue/MarkingDelegate.hpp]: If there is packet overflow in the example we will not handle it? What is this method supposed to do?

Comment (ktbriggs) [inline example/glue/MarkingDelegate.hpp]: It handles language-specific aspects of packet overflow, main handling is done in OMR core. Example code has no special needs in this context.

**Comment by ktbriggs** [comment]
3542e89 moves OMR write barrier out of MM_Configuration into gc/base/standard/StandardWriteBarrier.hpp. This write barrier is used only by OMR example code.

Keeping these last 2 commits separate for review. Will squash into main commit when review is complete.

**Comment by charliegracie** [comment]
I reviewed the code and it is ready to be squashed

**Comment by ktbriggs** [comment]
Squashed and rebased.

**Pull request merged**